ykshim_client/src/lib.rs

```diff
@@ -0,0 +1,245 @@
+    }
+}
+
+pub struct CompiledTrace<I> {
+    compiled: *mut c_void,
+    _marker: PhantomData<I>,
+}
+
+unsafe impl<I> Send for CompiledTrace<I> {}
+unsafe impl<I> Sync for CompiledTrace<I> {}
+
+pub fn compile_trace<T>(mut sir_trace: SirTrace) -> Result<CompiledTrace<T>, CString> {
+    let mut err_msg = std::ptr::null_mut();
+    let compiled = unsafe { __ykshim_compile_trace(sir_trace.0, &mut err_msg) };
+    sir_trace.0 = ptr::null_mut(); // consumed.
+    if compiled.is_null() {
+        return Err(unsafe { CString::from_raw(err_msg) });
+    }
+    Ok(CompiledTrace {
+        compiled,
+        _marker: PhantomData,
+    })
+}
+
+impl<I> CompiledTrace<I> {
+    pub fn ptr(&self) -> *const u8 {
+        unsafe { __ykshim_compiled_trace_get_ptr(self.compiled) as *const u8 }
+    }
+
+    /// Execute the trace with the given interpreter context.
+    pub unsafe fn execute(&self, icx: &mut I) -> bool {
+        let f = mem::transmute::<_, fn(&mut I) -> bool>(self.ptr());
+        f(icx)
+    }
+}
+
+impl<I> Drop for CompiledTrace<I> {
+    fn drop(&mut self) {
+        unsafe { __ykshim_compiled_trace_drop(self.compiled) }
+    }
+}
+
+pub fn sir_body_ret_ty(sym: &str) -> TypeId {
+    let sym_c = CString::new(sym).unwrap();
+    let mut cgu = CguHash(0);
+    let mut idx = 0;
+    unsafe { __ykshim_body_ret_ty(sym_c.into_raw(), &mut cgu, &mut idx) };
+    (cgu, idx)
+}
+
+pub struct LocalDecl {
+    pub ty: TypeId,
+    pub referenced: bool,
+}
+
+impl LocalDecl {
+    pub fn new(ty: TypeId, referenced: bool) -> Self {
+        Self { ty, referenced }
+    }
+}
+
+pub struct TraceCompiler(*mut c_void);
+
+impl TraceCompiler {
+    pub fn new(local_decls: HashMap<Local, LocalDecl>) -> Self {
+        let tc = unsafe { __ykshim_tracecompiler_default() };
+        for (l, decl) in local_decls.iter() {
+            unsafe {
+                __ykshim_tracecompiler_insert_decl(tc, *l, decl.ty.0, decl.ty.1, decl.referenced)
+            };
+        }
+
+        Self(tc)
+    }
+
+    pub fn local_to_location_str(&mut self, local: Local) -> String {
+        let ptr = unsafe { __ykshim_tracecompiler_local_to_location_str(self.0, local.0) };
+        String::from(unsafe { CString::from_raw(ptr).to_str().unwrap() })
+    }
+
+    pub fn local_dead(&mut self, local: Local) {
+        unsafe { __ykshim_tracecompiler_local_dead(self.0, local.0) };
+    }
+
+    pub fn find_symbol(sym: &str) -> *mut c_void {
+        let ptr = CString::new(sym).unwrap().into_raw();
+        unsafe { __ykshim_tracecompiler_find_sym(ptr) }
+    }
+}
+
+impl Drop for TraceCompiler {
+    fn drop(&mut self) {
+        unsafe { __ykshim_tracecompiler_drop(self.0) }
+    }
+}
+
+pub fn interpret_body<I>(body_name: &str, icx: &mut I) {
+    let body_cstr = CString::new(body_name).unwrap();
+    unsafe { __yktest_interpret_body(body_cstr.into_raw(), icx as *mut _ as *mut u8) };
+}
+
+pub fn reg_pool_size() -> usize {
+    unsafe { __yktest_reg_pool_size() }
+}
```

Comment:
It may be a good idea to put the wrappers for the test api in a separate module.

Reply:
Yes, I was thinking in a follow up PR.